### PR TITLE
Fixed #29008 -- Fixed crash of 404 debug page when URL path converter raises Http404.

### PR DESCRIPTION
**Pull request by nsiregar**
Not sure while writing test for this, since `Resolver404` is instance of `Http404` and while raised it only pass.

### CODE REVIEW

**Comment by felixxm** [comment]
@nsiregar Thanks for this patch :+1: I'm not sure if it is a proper solution I would rather catch `Http404` errors raised by [`converter.to_python()`](https://github.com/django/django/blob/5b57798513066bb6d5ea54225d0bd030a089c501/django/urls/resolvers.py#L261) and update [documentation](https://docs.djangoproject.com/en/2.2/topics/http/urls/#registering-custom-path-converters), because that's the main issue. Wrong technical page is only a consequence.

**Comment by nsiregar** [comment]
@felixxm thank you for reviewing, working on this

File: django/urls/resolvers.py [hunk 1]
```diff
@@ -258,7 +259,7 @@ def match(self, path):
                 converter = self.converters[key]
                 try:
                     kwargs[key] = converter.to_python(value)
-                except ValueError:
+                except Http404:
```
Comment (felixxm) [inline django/urls/resolvers.py]: We shouldn't change the current behavior because it is backward incompatible and it wasn't my intention. IMO we should catch both:
```python
except (Http404, ValueError):
```

File: django/urls/converters.py [hunk 2]
```diff
@@ -26,7 +31,10 @@ class UUIDConverter:
     regex = '[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}'
 
     def to_python(self, value):
-        return uuid.UUID(value)
+        try:
+            return uuid.UUID(value)
+        except ValueError:
+            raise Http404()
```
Comment (felixxm) [inline django/urls/converters.py]: Please revert this change.

File: django/urls/converters.py [hunk 3]
```diff
@@ -1,12 +1,17 @@
...
 
 class IntConverter:
     regex = '[0-9]+'
 
     def to_python(self, value):
-        return int(value)
+        try:
+            return int(value)
+        except ValueError:
+            raise Http404()
```
Comment (felixxm) [inline django/urls/converters.py]: Please revert this change.

File: docs/topics/http/urls.txt [hunk 4]
```diff
@@ -146,20 +146,25 @@ A converter is a class that includes the following:
 
 * A ``to_python(self, value)`` method, which handles converting the matched
   string into the type that should be passed to the view function. It should
-  raise ``ValueError`` if it can't convert the given value. A ``ValueError`` is
-  interpreted as no match and as a consequence a 404 response is sent to the
+  raise ``Http404`` if it can't convert the given value. A ``Http404`` is
+  raised as no match and as a consequence a 404 response is sent to the
```
Comment (felixxm) [inline docs/topics/http/urls.txt]:
```
raise ``ValueError`` or ``Http404`` if it can't convert the given value. A
``ValueError`` is interpreted as no match and as a consequence a 404
response is sent to the user.
```

File: docs/topics/http/urls.txt [hunk 5]
```diff
@@ -146,20 +146,25 @@ A converter is a class that includes the following:
...
+
     class FourDigitYearConverter:
         regex = '[0-9]{4}'
 
         def to_python(self, value):
-            return int(value)
+            try:
+                return int(value)
+            except ValueError:
+                raise Http404()
```
Comment (felixxm) [inline docs/topics/http/urls.txt]: Please revert this change.

File: tests/urlpatterns/tests.py [hunk 6]
```diff
@@ -208,7 +209,7 @@ class ConversionExceptionTests(SimpleTestCase):
     def test_resolve_value_error_means_no_match(self):
         @DynamicConverter.register_to_python
         def raises_value_error(value):
-            raise ValueError()
+            raise Http404()
```
Comment (felixxm) [inline tests/urlpatterns/tests.py]: Please add a second test for converter that will raise `Http404`, also a test for technical response (in `view_tests/tests/test_debug.py`) when used converter returns `Http404`.

**Comment by felixxm** [comment]
@nsiregar After reconsideration I think that the first version of your PR with catching `Http404` instead of `Resolver404` was correct, sorry. I'll restore it.
